Parres pr dgemm nstream

scoumeri opened the pull request:
- Relates to #1933 [bug report/feature request/story] from github issues
- Fixes #1921 change request from github issues.

[Brief description of the change]

Adding ParRes Kernel (specifically dgemm and nstream) to the integration environment for use with gpus.

lhlawson committed:
Added ParRes workload hard coded build script

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Added patch and build info for nstream-cublas for mpi

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

lhlawson committed:
Added untested MPI support for transpose and sgemm cublas

Signed-off-by: lowren.h.lawson@intel.com <lowren.h.lawson@intel.com>

scoumeri committed:
Merge remote-tracking branch 'lhlawson/public-lhlawson-parres' into parres_pr_a

scoumeri committed:
* Pulls in public-lhlawson-parres branch

 * Updated build.sh script to remove some hardcoded paths

 * created parres.py for running dgemm and nstream ParRes kernels

 * created experiment/monitor/run_monitor_parres_dgemm.py and run_monitor_nstream.py

Signed-off-by: sari.coumeri@intel.com <sari.coumeri@intel.com>

scoumeri committed:
Removed sgemm and transpose from build script and Makefile as this PR is only for dgemm and nstream

Signed-off-by: sari.coumeri@intel.com <sari.coumeri@intel.com>

lhlawson commented on integration/apps/parres/Makefile.mk:
```diff
@@ -0,0 +1,38 @@
+	      integration/apps/parres/__init__.py \
+              integration/apps/parres/build.sh \
+              integration/apps/parres/README.md \
+	      integration/apps/parres/parres.py \
```
wsp

scoumeri replied:
Fixed this

cmcantalupo commented on integration/apps/parres/parres.py:
```diff
@@ -0,0 +1,321 @@
+
+        elif mpi_pin_default:
+            affinity_disable = True
+            mpi_pin_list = "1,2,3,21,22"
```
This data seems like it is architecture specific, can it be derived from machine file?

scoumeri replied:
This is a remnant from the Hacc script (which I wrote first but will PR later).  That had a large affinity mapping which was painful to specify (but didn't actually seem to make a difference on saruman). I just left the capability in for dgemm and nstream, but don't necessarily need to use it.  

I can get rid of the mpi_pin_default.  I can also get rid of mpi_pin_list and make the user set the env variable outside of the run.  I didn't automatically set the geopm-affinity-disable, so that right now has to happen on the command line anyway.

Lowren, any preference?

scoumeri replied:
Ok, discussed with Lowren.  I'll drop the options.

scoumeri replied:
I've updated the code to drop the pinning options.

cmcantalupo commented on integration/apps/parres/parres.py:
```diff
@@ -0,0 +1,321 @@
+
+        elif mpi_pin_default:
+            affinity_disable = True
+            mpi_pin_list = "1,21"
```
Derive from machine file?

scoumeri replied:
Same as above comment.

scoumeri replied:
I've updated the code to drop the pinning options.

cmcantalupo commented on integration/apps/parres/parres.py:
```diff
@@ -0,0 +1,321 @@
+    def get_bash_setup_commands(self):
+        setup_commands = textwrap.dedent('''
+        ulimit -s unlimited
+        export I_MPI_FABRICS=shm
```
Is this for single node only runs? If so, is this enforced somewhere?

scoumeri replied:
This has all been run on saruman which is 1 node.  According to Lowren: As is, dgemm may work for multiple nodes, but nstream probably will only work for single node.  

Will keep it to single node and I'll add a check to enforce this.

scoumeri replied:
Added a check to make sure it's running on a single node.

scoumeri replied:
Trying to resolve conversation.

cmcantalupo reviewed (commented):
Thanks for the contributions.  Couple of questions about the affinity settings and the intel specific MPI fabric setting.

scoumeri committed:
* Removed pinning argument

 * Added check that only 1 node is being used

Signed-off-by: Sari Coumeri <sari.coumeri@intel.com>

lhlawson commented on integration/apps/parres/parres.experiment.setup:
```diff
@@ -0,0 +1,37 @@
+#clears clock settings
+nvidia-smi -rgc
+#sets power limit
+nvidia-smi -pl 300 
```
RAC is a user level control only if root allows it.  
RGC and PL are root level controls.  May be worthwhile to call this out.

scoumeri replied:
Added comment in file about needing root level access.

lhlawson commented on integration/apps/parres/parres.py:
```diff
@@ -0,0 +1,296 @@
+            os.chmod(self._parres_teardown, 0o755)
+            subprocess.call(self._parres_teardown, shell=True)
+
+
```
```suggestion
def parse_fom(self, log_path):
        result = None
        key = 'AVG Rate (MF/s): '
        with open(log_path) as fid:
            for line in fid.readlines():
                if key in line:
                    result = line.split(': ')[1]
                    result = float(result.split(' ')[0])
                    break
        return result
```

lhlawson replied:
Suggestion to add figure of merit.  Untested.

scoumeri replied:
Added FOM code.

lhlawson commented on integration/apps/parres/parres.py:
```diff
@@ -0,0 +1,296 @@
+            os.chmod(self._parres_teardown, 0o755)
+            subprocess.call(self._parres_teardown, shell=True)
+
+
```
```suggestion
```suggestion
def parse_fom(self, log_path):
        result = None
        key = 'Rate (MB/s): '
        with open(log_path) as fid:
            for line in fid.readlines():
                if key in line:
                    result = line.split(': ')[1]
                    result = float(result.split(' ')[0])
                    break
        return result
```
```

scoumeri replied:
Added FOM code.

scoumeri committed:
- Updated dgemm and nstream to have a FOM
 - Added comment to parres.experiment.setup that commands require root access
 - Updated nstream to the cuda version which has better performance than the cublas version

Signed-off-by: Sari Coumeri <sari.coumeri@intel.com>

scoumeri commented:
Made additional change to switch nstream from cublas to cuda as it seems to have better performance.

lhlawson commented:
> 
> 
> Made additional change to switch nstream from cublas to cuda as it seems to have better performance.

LGTM, using the patch to add a new file should work (could do a cp & modify patch instead).

cmcantalupo reviewed (approved):
Thanks @scoumeri
